C++: AST dataflow through smart pointers

MathiasVP opened the pull request:
Part of https://github.com/github/codeql-c-analysis-team/issues/255.

This PR adds AST dataflow through smart pointers using a technique similar to the one we used for iterators in https://github.com/github/codeql/pull/4135.

CPP-differences: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1878/

This PR is just focused on getting data "out of" functions that take a smart pointer, similar to how `DefinitionByReference` works for raw pointers. The actual handling of read and store steps is done in https://github.com/github/codeql/pull/5607.

I recommend a commit-by-commit review of the first three commits. The <s>final</s> commit (32a8b9a857eb66a5b471de771dff380a61654e7b) just moves lines around.

intrigus-lgtm commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll:
```diff
@@ -243,6 +244,54 @@ private module PartialDefinitions {
+        innerDefinedExpr = getAPointerWrapperAccess(pointer)
+      )
+      or
+      // iterators passed by value without a copy constructor
```
I don't think this is an iterator :D
Also see the other comment below.

MathiasVP replied:
Fixed in ecbce88ec78710f67c0160f7ae58e58c027fb73c. Thanks!

geoffw0 commented on cpp/ql/test/library-tests/dataflow/taint-tests/stl.h:
```diff
@@ -348,7 +348,7 @@ namespace std {
 	public:
 		shared_ptr() noexcept;
-		explicit shared_ptr(T*);
+		explicit shared_ptr(T*); shared_ptr(const shared_ptr&) noexcept;
```
This is a bit hard to read, is it a pain to put it on its own line?

MathiasVP replied:
The only pain point is the fact that it'll generate a large diff in the "Files changed" tab on GitHub. But I'll do it, and whoever ends up reviewing this can view the changes commit-by-commit to avoid inspecting the inflated diff.

MathiasVP replied:
Fixed in 32a8b9a857eb66a5b471de771dff380a61654e7b.

MathiasVP commented:
> CPP-differences: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1878/

CPP-differences looks fine. The difference in performance seems to be noise: the largest timing change is in a query that doesn't use dataflow.

jbj commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll:
```diff
@@ -243,6 +244,54 @@ private module PartialDefinitions {
+    Expr innerDefinedExpr;
+
+    SmartPointerPartialDefinition() {
+      exists(Expr convertedInner |
+        not this instanceof Conversion and
+        valueToUpdate(convertedInner, this.getFullyConverted(), node) and
+        innerDefinedExpr = convertedInner.getUnconverted() and
+        innerDefinedExpr = getAPointerWrapperAccess(pointer)
+      )
```
Is this case fully subsumed by the charpred of `VariablePartialDefinition`?

MathiasVP replied:
It is. This means that `*smart_ptr = e` is already a partial definition according to `VariablePartialDefinition`. The problem is that for `VariablePartialDefinition` it's a partial definition _of the variable_. Whereas for `SmartPointerPartialDefinition` it's a partial definition _of the call to `operator*`_. The same problem arises in `IteratorPartialDefinition`. I'm not exactly sure how to fix this short of making `valueToUpdate` more complex. To reduce duplication I've refactored the common parts out into a predicate in 0a6aef71a2f957386f895d594f227a0c67c387be.

jbj commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll:
```diff
@@ -243,6 +244,54 @@ private module PartialDefinitions {
   }
 
+  class SmartPointerPartialDefinition extends PartialDefinition {
+    Variable pointer;
```
I've lost track of how everything works in `PartialDefinition`, but I suspect it may be hurting more than helping to require a `Variable` here. It's no longer required in `VariablePartialDefinition`, despite the name, and it only seems to be used for the deprecated members.

MathiasVP replied:
It's used by `partiallyDefinesVariableAt`. This predicate is used by `BlockVar::definedPartiallyAt`, and it's behavior is important for the `Node -> FlowVar -> VariableAccess` case in `simpleLocalFlowStep`. Specifically, it's used to get flow out of a partial definition and into variable accesses and parameters. It might be possible to refactor this to something simpler, but I think it's a non-trivial change.

MathiasVP commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -330,6 +332,18 @@ class IteratorPartialDefinitionNode extends PartialDefinitionNode {
+ * A synthetic data flow node used for flow into a collection when a smart pointer
+ * write occurs in a callee.
+ */
+class SmartPointerPartialDefinitionNode extends PartialDefinitionNode {
```
Fixed in 0a6aef71a2f957386f895d594f227a0c67c387be.

MathiasVP commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll:
```diff
@@ -832,10 +893,19 @@ module FlowVar_internal {
+  Call getAPointerWrapperAccess(Variable pointer) {
+    pointer.getUnspecifiedType() instanceof PointerWrapper and
+    [result.getQualifier(), result.getAnArgument()] = pointer.getAnAccess()
+  }
```
Yeah, good point. The name is a result of me following the pattern from `getAnIteratorAccess` which has a more complicated job to do.

The predicate is used by the char pred of `SmartPointerPartialDefinition`. In the case of `operator->` on a smart pointer the `valueToUpdate` predicate on line 254:
```codeql
valueToUpdate(convertedInner, this.getFullyConverted(), node) and
```
follows the flow from `this.getFullyConverted()` down to the call to `operator->`, and the `getAPointerWrapperAccess` predicate then related that call to the variable being updated.

The problem with this predicate is really that it serves two purposes depending on which disjunction of the char pred we're in:
1. In the `valueToUpdate` case (which covers the case of `smart_ptr->x`) we want to return the qualifier, and
2. In the copy constructor case (which covers the case of passing a smart pointer by value to a function) we want to return the 0'th argument since copy constructors don't have a qualifier).

After writing down this explanation I realize that `getAPointerWrapperAccess` can be specialized more than what it's currently doing. I've fixed this in 0a6aef71a2f957386f895d594f227a0c67c387be (which hopefully should also make the goal of the predicate more clear).

jbj commented:
This PR is clearer to me after your latest round of comments, thanks.

I wonder if I can think of this PR as ensuring that an update through `smartPtr.operator->()` is seen as an update through `smartPtr`. If that's the case, it seems similar to how we ensure that updates through `rawPtr + 1` are seen as updates through `rawPtr`. That happens [in `AddressFlow.qll`](https://github.com/github/codeql/blob/652e8b4872a1f17389d846533331ab1ea5900f7d/cpp/ql/src/semmle/code/cpp/dataflow/internal/AddressFlow.qll#L69-L93), where there are rules for how a pointer to "essentially the same thing" is constructed from an existing pointer. Instead of changing `FlowVar.qll`, perhaps a more general solution is to add rules to `AddressFlow.qll` for `operator->` and copy constructors of smart pointers.

MathiasVP commented:
> I wonder if I can think of this PR as ensuring that an update through `smartPtr.operator->()` is seen as an update through `smartPtr`. If that's the case, it seems similar to how we ensure that updates through `rawPtr + 1` are seen as updates through `rawPtr`. That happens [in `AddressFlow.qll`](https://github.com/github/codeql/blob/652e8b4872a1f17389d846533331ab1ea5900f7d/cpp/ql/src/semmle/code/cpp/dataflow/internal/AddressFlow.qll#L69-L93), where there are rules for how a pointer to "essentially the same thing" is constructed from an existing pointer. Instead of changing `FlowVar.qll`, perhaps a more general solution is to add rules to `AddressFlow.qll` for `operator->` and copy constructors of smart pointers.

You were right! 2329b3160105fd2b7bfc34af13563f740ae41df9 replaces the new partial definition with additional steps in `AddressFlow` that gives the same flow. 🎉

jbj commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/AddressFlow.qll:
```diff
@@ -218,7 +247,9 @@ private predicate referenceToUpdate(Expr reference, Expr outer, ControlFlowNode
     exists(ReferenceType rt | rt = outer.getType().stripTopLevelSpecifiers() |
-      not rt.getBaseType().isConst()
+      not rt.getBaseType().isConst() or
+      rt.getBaseType().getUnspecifiedType() =
+        any(PointerWrapper wrapper | not wrapper.pointsToConst())
```
In this new rule, `rt` is something like `const shared_ptr<int> &`: there are two levels of indirection before the `int`. Therefore I'd expect this rule to be in a new "reference-to-pointer to update" predicate.

MathiasVP replied:
Fixed in 5aeaab7c6d7ff786c970b41f1bebce8c6040d838.

jbj reviewed (commented):
The new results look great, but I'm concerned about getting those results by conflating "reference to (smart) pointer to T" with "reference to T". If you think this conflation is safe, I'll have a go at creating some test cases to see if it might cause anything to go wrong.

MathiasVP commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/AddressFlow.qll:
```diff
@@ -58,6 +59,8 @@ private predicate pointerToLvalueStep(Expr pointerIn, Expr lvalueOut) {
   pointerIn = lvalueOut.(ArrayExpr).getArrayBase().getFullyConverted()
   or
   pointerIn = lvalueOut.(PointerDereferenceExpr).getOperand().getFullyConverted()
+  or
+  pointerIn = lvalueOut.(OverloadedPointerDereferenceExpr).getQualifier().getFullyConverted()
```
Good point about this rule belonging in `pointerToReferenceStep`. This rule isn't actually needed when we have the rule I added in `lvalueToReferenceStep`.

MathiasVP replied:
5aeaab7c6d7ff786c970b41f1bebce8c6040d838 deletes this rule in and adds the version from `lvalueToReferenceStep` to a new `pointerToReferenceStep` predicate.

MathiasVP commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/AddressFlow.qll:
```diff
@@ -106,6 +115,13 @@ private predicate referenceToPointerStep(Expr referenceIn, Expr pointerOut) {
+    call.getTarget() = copy and
+    // The 0'th argument is the value being copied.
+    referenceIn = call.getArgument(0).getFullyConverted()
+  )
```
Fixed in 5aeaab7c6d7ff786c970b41f1bebce8c6040d838.

MathiasVP commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/AddressFlow.qll:
```diff
@@ -94,6 +97,12 @@ private predicate pointerToPointerStep(Expr pointerIn, Expr pointerOut) {
+    referenceOut.getUnspecifiedType() instanceof ReferenceType and
+    call = wrapper.getAnUnwrapperFunction().getACallToThisFunction() and
+    lvalueIn = call.getQualifier().getFullyConverted()
+  )
```
Good point. Fixed in 5aeaab7c6d7ff786c970b41f1bebce8c6040d838.

jbj commented on cpp/ql/src/semmle/code/cpp/dataflow/internal/AddressFlow.qll:
```diff
@@ -300,6 +383,21 @@ private predicate referenceFromVariableAccess(VariableAccess va, Expr reference)
+private predicate pointerToReferenceToPointerStep(Expr pointerIn, Expr referenceToPointerOut) {
+  pointerIn.getConversion() = referenceToPointerOut.(ReferenceToExpr)
+  or
+  exists(PointerWrapper wrapper, Call call | call = referenceToPointerOut |
+    referenceToPointerOut.getUnspecifiedType() instanceof ReferenceType and
+    call = wrapper.getAnUnwrapperFunction().getACallToThisFunction() and
+    pointerIn = call.getQualifier().getFullyConverted()
+  )
```
This is the same code as `pointerToReferenceStep`. I don't think it belongs here.

jbj reviewed (commented):
Some comments written during our Zoom call. Conclusion: run CPP-Differences without the last commit; if that looks good, drop the last commit, and then we can merge.

MathiasVP commented:
> Some comments written during our Zoom call. Conclusion: run CPP-Differences without the last commit; if that looks good, drop the last commit, and then we can merge.

CPP-difference without the last commit: https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1908/

MathiasVP commented:
@jbj I've reverted the last commit (and fixed up the comments from the review of 2329b3160105fd2b7bfc34af13563f740ae41df9). I believe this PR is good to go now. 🤞
